node/bindnode/repr.go

```diff
@@ -53,7 +53,11 @@ func (w *_nodeRepr) Kind() datamodel.Kind {
 		return datamodel.Kind_Map
 	case schema.UnionRepresentation_Kinded:
 		haveIdx, _ := unionMember(w.val)
-		mtyp := w.schemaType.(*schema.TypeUnion).Members()[haveIdx] // TODO this is fragile: panics with index-out-of-range if the user has nil'd all fields.
+		if haveIdx < 0 {
+			// Invalid state; no union member is set.
+			return datamodel.Kind_Invalid
+		}
```

Comment:
I think this _could_ panic, fwiw, if it likes.  It should just say something explicit about "bindnode: node has invalid state (did you modify it directly, incorrectly?)".

I've tried really hard to avoid using `datamodel.Kind_Invalid` almost anywhere at all (the one exception being [schema.TypeUnion.RepresentationBehavior()](https://github.com/ipld/go-ipld-prime/blob/3d82a69fa632144c64f7e205f3b0f2ee28d27c27/schema/typeMethods.go#L40), because it has literally no other choice)... so I think I might actually prefer a panic here over pushing around invalid states to blow up somewhere else that will end up with less information about why.

Also: I think I used one-indexed numbering for union members in the gengo stuff in order to make sure uninitialized zero values would stand out like a sore thumb (as they should).  Might be worth doing here too.

Reply:
Agreed. I actually added this because I ran into the panic while adding test cases. In particular, because AssignBytes would fail to do anything, so I'd end up with an invalid union state.

After this change, the dag-json encode gives an "encountered absent" error, which is slightly misleading at best. A clear panic seems better.